File: elir_line_control/scripts/auto_move.py
#!/usr/bin/env python
import rospy
from std_msgs.msg import Float64
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Float32
from math import pi,pow,sqrt

class Robot:
    def __init__(self):
        #initiate the note called auto_move_robot with a random number associated
        rospy.init_node('auto_move_robot', anonymous=True)

        #create the topic for shaft 1c
        self.traction_1c_path = rospy.Publisher ('/robot/traction_1c_controller/command', Float64, queue_size=10)

        #create the topic for shaft 1f
        self.traction_1f_path = rospy.Publisher ('/robot/traction_1f_controller/command', Float64, queue_size=10)

        #create the topic for shaft 2f
        self.traction_2f_path = rospy.Publisher ('/robot/traction_2f_controller/command', Float64, queue_size=10)

        #create the topic for shaft 1b
        self.traction_1b_path = rospy.Publisher ('/robot/traction_1b_controller/command', Float64, queue_size=10)

        #create the topic for shaft 2b
        self.traction_2b_path = rospy.Publisher ('/robot/traction_2b_controller/command', Float64, queue_size=10)



        self.joint_1f_path = rospy.Publisher ('/robot/joint_1f_controller/command', Float64, queue_size=10)

        self.joint_1b_path = rospy.Publisher ('/robot/joint_1b_controller/command', Float64, queue_size=10)

        self.joint_2f_path = rospy.Publisher ('/robot/joint_2f_controller/command', Float64, queue_size=10)

        self.joint_2b_path = rospy.Publisher ('/robot/joint_2b_controller/command', Float64, queue_size=10)

        self.claw_1f_path = rospy.Publisher ('/robot/claw_1f_controller/command', Float64, queue_size=10)
        self.claw_2f_path = rospy.Publisher ('/robot/claw_2f_controller/command', Float64, queue_size=10)
        self.claw_1b_path = rospy.Publisher ('/robot/claw_1b_controller/command', Float64, queue_size=10)
        self.claw_2b_path = rospy.Publisher ('/robot/claw_2b_controller/command', Float64, queue_size=10)
        self.claw_1c_path = rospy.Publisher ('/robot/claw_1c_controller/command', Float64, queue_size=10)


        self.hokuyo_path = rospy.Subscriber ('/robot/scan', LaserScan,    self.callback_hokuyo)
        #variable that contains the speed of the robot
        self.speed = Float64()
        #Initialize the speed as 2
        self.speed = 15

        #Variable that contains the duration of movement and initializa as 3 secondes
        self.time = 3
        self.hokuyu_info = Float32()
        self.hojuyu_inc_angle = Float32()
        #Go to panel for decision making
        self.IHM_Panel()

    def IHM_Panel(self):
        print "\nCurrent speed:", self.speed
        print "Duration of movement in sec:", self.time
        print "current time", rospy.Time.now().to_sec()
        print "\nTo move forward press 1"
        print "To move backwards press 2"
        print "To change the speed press 3"
        print "To change the duration of the movement press 4"
        print "To move until find, approx. and surpass an obstacle press 5"
        print "To quit press 6"
        choice = input("\nInsert the number here: ")
        if choice == 1:
            self.move_forward()
        if choice == 2:
            self.move_backward()
        if choice == 3:
            self.change_speed()
        if choice == 4:
            self.change_time()
        if choice == 5:
            self.detect_obs()
        if choice == 6:
            sys.exit(1)
        # if choice == None:
        #     print "Inserted an invalid number, going back to panel"
        #     self.IHM_Panel()
        else:
            print "Inserted an invalid number, going back to panel"
            self.IHM_Panel()

    def move_forward(self):
        print "\nMoving Forward!"
        self.traction_1c_path.publish(self.speed)
        self.traction_1f_path.publish(-self.speed)
        self.traction_2f_path.publish(self.speed)
        self.traction_1b_path.publish(-self.speed)
        self.traction_2b_path.publish(self.speed)
        current_time = 0
        starting_time = rospy.get_time()
        # print starting_time
        # x = 0
        while current_time < self.time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
            # if x == 0:
            #     print "actual time:", actual_time
            #     print "current time:", current_time
            #     x = 1
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        print "Moving back to panel"
        self.IHM_Panel()

    def move_backward(self):
        print "\nMoving backwards!"
        self.traction_1c_path.publish(-self.speed)
        self.traction_1f_path.publish(self.speed)
        self.traction_2f_path.publish(-self.speed)
        self.traction_1b_path.publish(self.speed)
        self.traction_2b_path.publish(-self.speed)
        current_time = 0
        starting_time = rospy.Time.now().to_sec()
        while current_time < self.time:
            actual_time = rospy.Time.now().to_sec()
            current_time = actual_time - starting_time
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        print "Moving back to panel"
        self.IHM_Panel()

    def change_speed(self):
        self.speed = abs(input("\nInsert new absolute value of the speed:"))
        print"Moving back to panel"
        self.IHM_Panel()

    def change_time(self):
        self.time = abs(input("\nInsert new duration time of movement:"))
        print"Moving back to panel"
        self.IHM_Panel()

    def callback_hokuyo(self, data):
        self.hokuyu_info = filter(lambda x: x != float('Inf'), data.ranges)
        self.hojuyu_inc_angle = data.angle_increment

    # def hokuyo_OLDER_VERSIONS(self):
    #     largura = 0
    #     while largura*1000 < 30:
    #         print "Obstacle not detected, proceding!"
    #         length = len(self.hokuyu_info)
    #         pos_ini = Float32()
    #         pos_final = Float32()
    #         lock = 1
    #         soma = 0
    #         for pos in range(0,(length)):
    #             if self.hokuyu_info[pos] != float('Inf'):
    #                 soma = soma + self.hokuyu_info[pos]
    #                 if lock == 1:
    #                     pos_ini = pos
    #                     lock = 0
    #             else:
    #                 if lock == 0:
    #                     pos_final = pos - 1
    #                     break
    #         media = soma/(pos_final-pos_ini+1)
    #         largura = media*self.hojuyu_inc_angle* ((pos_final-pos_ini)+1)
    #         print "largura lida em mm:", largura*1000
    #         self.traction_1c_path.publish(self.speed)
    #         self.traction_1f_path.publish(self.speed)
    #         self.traction_2f_path.publish(self.speed)
    #         self.traction_1b_path.publish(self.speed)
    #         self.traction_2b_path.publish(self.speed)
    #     print "Obstacle detected, preparing to aprox. it"
    #     self.traction_1c_path.publish(0)
    #     self.traction_1f_path.publish(0)
    #     self.traction_2f_path.publish(0)
    #     self.traction_1b_path.publish(0)
    #     self.traction_2b_path.publish(0)
    #     self.aprox_obs(sqrt(pow(media,2)-pow(0.4,2)))
    #     print"Moving back to panel"
    #     self.IHM_Panel()

    def detect_obs(self):
        largura = 0
        while largura*1000 < 30:
            print "Obstacle not detected, proceding!"
            soma = 0
            hokuyu_info = self.hokuyu_info
            length = len(hokuyu_info)
            for pos in range(0,(length)):
                soma = soma + hokuyu_info[pos]
            largura = self.hojuyu_inc_angle*soma
            print "largura lida em mm:", largura*1000
            self.traction_1c_path.publish(self.speed)
            self.traction_1f_path.publish(-self.speed)
            self.traction_2f_path.publish(self.speed)
            self.traction_1b_path.publish(-self.speed)
            self.traction_2b_path.publish(self.speed)
        print "Obstacle detected, preparing to aprox. it"
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        self.aprox_obs(sqrt(pow(soma/length,2)-pow(0.4,2)))
        print"Finished all movements, Moving back to panel"
        self.IHM_Panel()

    def aprox_obs(self,dist):
        speed = 5
        time = ((dist-0.08)/speed)*60
        self.traction_1c_path.publish(speed)
        self.traction_1f_path.publish(-speed)
        self.traction_2f_path.publish(speed)
        self.traction_1b_path.publish(-speed)
        self.traction_2b_path.publish(speed)
        current_time = 0
        starting_time = rospy.get_time()
        # print starting_time
        # x = 0
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
            # if x == 0:
            #     print "actual time:", actual_time
            #     print "current time:", current_time
            #     x = 1
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        print "aprrox. completed, preparing to surpass it"
        self.surpass_obs()

    def surpass_obs(self):
        #move up front arm to allow claws to rotate
        self.joint_1f_path.publish(-0.3)
        self.joint_2f_path.publish(0.1)
        #Create a delay, to allow the arm to go up before rotating claws
        time = 1
        current_time = 0
        starting_time = rospy.get_time()
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        #Rotate claws
        self.claw_1f_path.publish(3.14)
        self.claw_2f_path.publish(3.14)
        #Move some distance to allow front arm to pass the obstacle
        dist = 0.6
        speed = 5
        time = ((dist-0.08)/speed)*60
        current_time = 0
        starting_time = rospy.get_time()
        self.traction_1c_path.publish(speed)
        self.traction_1f_path.publish(-speed)
        self.traction_2f_path.publish(speed)
        self.traction_1b_path.publish(-speed)
        self.traction_2b_path.publish(speed)
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        #rotate the claws back to original position
        self.claw_1f_path.publish(0)
        self.claw_2f_path.publish(0)
        #small delay to give sufficient time to claws rotate back to position
        time = 0.5
        current_time = 0
        starting_time = rospy.get_time()
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        #move front arm back to original position
        self.joint_1f_path.publish(0)
        self.joint_2f_path.publish(0)
        #move both arms down to force the center up
        self.joint_1f_path.publish(1.3)
        self.joint_1b_path.publish(1.3)
        #Create a delay, to allow the center to go up before rotating claw
        time = 1
        current_time = 0
        starting_time = rospy.get_time()
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        #Rotate center claw
        self.claw_1c_path.publish(3.14)
        #Move some distance to allow the center to pass the obstacle
        dist = 0.6
        speed = 5
        time = ((dist-0.08)/speed)*60
        current_time = 0
        starting_time = rospy.get_time()
        self.traction_1c_path.publish(speed)
        self.traction_1f_path.publish(-speed)
        self.traction_2f_path.publish(speed)
        self.traction_1b_path.publish(-speed)
        self.traction_2b_path.publish(speed)
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        #Rotate the center claw back to orginal position and arms back to positio to lower the center
        self.claw_1c_path.publish(0)
        self.joint_1f_path.publish(0)
        self.joint_1b_path.publish(0)
        #move up back arm to allow claws to rotate
        self.joint_1b_path.publish(-0.3)
        self.joint_2b_path.publish(0.1)
        #Create a delay, to allow the arm to go up before rotating claws
        time = 1
        current_time = 0
        starting_time = rospy.get_time()
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        #Rotate claws
        self.claw_1b_path.publish(3.14)
        self.claw_2b_path.publish(3.14)
        #Move some distance to allow front arm to pass the obstacle
        dist = 0.6
        speed = 5
        time = ((dist-0.08)/speed)*60
        current_time = 0
        starting_time = rospy.get_time()
        self.traction_1c_path.publish(speed)
        self.traction_1f_path.publish(-speed)
        self.traction_2f_path.publish(speed)
        self.traction_1b_path.publish(-speed)
        self.traction_2b_path.publish(speed)
        while current_time < time:
            actual_time = rospy.get_time()
            current_time = actual_time - starting_time
        self.traction_1c_path.publish(0)
        self.traction_1f_path.publish(0)
        self.traction_2f_path.publish(0)
        self.traction_1b_path.publish(0)
        self.traction_2b_path.publish(0)
        #move claws and arm back to original position
        self.claw_1b_path.publish(0)
        self.claw_2b_path.publish(0)
        self.joint_1b_path.publish(0)
        self.joint_2b_path.publish(0)
        print "Surpassed"

if __name__ == '__main__':
        #Creates a object called turtle10 of the class Turtle
        piro3 = Robot()

File: elir_line_control/scripts/auto_move.cpp
// my first program in C++
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <std_msgs/Float32.h>
#include <sensor_msgs/LaserScan.h>
#include <iostream>
#include <math.h>
#include <vector>
#include <ros/callback_queue.h>

class Robot {
    std_msgs::Float64 speed;
    std_msgs::Float64 speed_neg;
    std::vector<float> hokuyo_info;
    std::vector<float> hokuyo_info_no_treatment;
    std_msgs::Float32 hokuyo_inc_angle;
    std_msgs::Float64 stop_speed;


    ros::Publisher traction_1c_path, traction_1f_path, traction_2f_path, traction_1b_path, traction_2b_path;

    ros::Publisher joint_1f_path, joint_1b_path, joint_2f_path, joint_2b_path;

    ros::Publisher claw_1f_path, claw_2f_path, claw_1b_path, claw_2b_path, claw_1c_path;

    ros::Subscriber hokuyo_path;

    ros::NodeHandle nh;

    double dur_timer;
  public:
    Robot (void);
    void IHM_Panel(void);
    void move_forward(void);
    void move_backward(void);
    void change_speed(void);
    void change_time(void);
    void detect_obs(void);
    void aprox_obs(float dist);
    void surpass_obs(void);
    void callback_hokuyo(const sensor_msgs::LaserScan::ConstPtr& data);
};

Robot::Robot(void) {
  speed.data = 15;
  speed_neg.data = -speed.data;
  dur_timer = 3;
  hokuyo_inc_angle.data = 7;
  stop_speed.data = 0;

  traction_1c_path = nh.advertise<std_msgs::Float64>("/robot/traction_1c_controller/command", 1000);
  traction_1f_path = nh.advertise<std_msgs::Float64>("/robot/traction_1f_controller/command", 1000);
  traction_2f_path = nh.advertise<std_msgs::Float64>("/robot/traction_2f_controller/command", 1000);
  traction_1b_path = nh.advertise<std_msgs::Float64>("/robot/traction_1b_controller/command", 1000);
  traction_2b_path = nh.advertise<std_msgs::Float64>("/robot/traction_2b_controller/command", 1000);

  joint_1f_path = nh.advertise<std_msgs::Float64>("/robot/joint_1f_controller/command", 1000);
  joint_1b_path = nh.advertise<std_msgs::Float64>("/robot/joint_1b_controller/command", 1000);
  joint_2f_path = nh.advertise<std_msgs::Float64>("/robot/joint_2f_controller/command", 1000);
  joint_2b_path = nh.advertise<std_msgs::Float64>("/robot/joint_2b_controller/command", 1000);

  claw_1f_path = nh.advertise<std_msgs::Float64>("/robot/claw_1f_controller/command", 1000);
  claw_2f_path = nh.advertise<std_msgs::Float64>("/robot/claw_2f_controller/command", 1000);
  claw_1b_path = nh.advertise<std_msgs::Float64>("/robot/claw_1b_controller/command", 1000);
  claw_2b_path = nh.advertise<std_msgs::Float64>("/robot/claw_2b_controller/command", 1000);
  claw_1c_path = nh.advertise<std_msgs::Float64>("/robot/claw_1c_controller/command", 1000);

  hokuyo_path = nh.subscribe<sensor_msgs::LaserScan>("/robot/scan", 1000, &Robot::callback_hokuyo,this);

  Robot::IHM_Panel();
}

void Robot::IHM_Panel(void) {
  int choice;
  // ROS_INFO("Current Speed: %.2f", speed.data);
  // ROS_INFO("Duration of movement(sec): %.2f", dur_timer);
  // ROS_INFO("Current time: %.2f", ros::Time::now().toSec());
  // ROS_INFO("To move forward press 1");
  // ROS_INFO("To move backwards press 2");
  // ROS_INFO("To change the speed press 3");
  // ROS_INFO("To change the duration of the movement press 4");
  // ROS_INFO("To move until find, approx. and surpass an obstacle press 5");
  // ROS_INFO("To quit press 6");
  // ROS_INFO("Insert the number:");

  std::cout << "\nCurrent Speed: " << speed.data << "\n";
  std::cout << "Duration of movement in sec: " << dur_timer << "\n";
  std::cout << "current time: " << ros::Time::now().toSec() << "\n\n";
  std::cout << "To move forward press 1" << "\n";
  std::cout << "To move backwards press 2" << "\n";
  std::cout << "To change the speed press 3" << "\n";
  std::cout << "To change the duration of the movement press 4" << "\n";
  std::cout << "To move until find, approx. and surpass an obstacle press 5" << "\n";
  std::cout << "To quit press 6" << "\n";
  std::cout << "Insert the number here: ";
  std::cin >> choice;
  std::cout << "\n";
  if (choice == 1) {
    Robot::move_forward();
  }
  else if (choice == 2) {
    Robot::move_backward();
  }
  else if (choice == 3) {
    Robot::change_speed();
  }
  else if (choice == 4) {
    Robot::change_time();
  }
  else if (choice == 5) {
    Robot::detect_obs();
  }
  else if (choice == 6) {
    exit(1);
  }
  else {
    std::cout << "There isn't this option, moving back to choice panel" << "\n";
    Robot::IHM_Panel();
  }
}

void Robot::move_forward(void) {
  std::cout << "Moving Forward!" <<"\n";
  double current_time, starting_time, actual_time;

  traction_1c_path.publish(speed);
  traction_1f_path.publish(speed_neg);
  traction_2f_path.publish(speed);
  traction_1b_path.publish(speed_neg);
  traction_2b_path.publish(speed);
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < dur_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }
  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);
  std::cout << "Moving back to panel" <<"\n";
  Robot::IHM_Panel();
}

void Robot::move_backward(void) {
  std::cout << "Moving Backwards!" <<"\n";
  double current_time, starting_time, actual_time;
  traction_1c_path.publish(speed_neg);
  traction_1f_path.publish(speed);
  traction_2f_path.publish(speed_neg);
  traction_1b_path.publish(speed);
  traction_2b_path.publish(speed_neg);
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < dur_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }
  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);
  std::cout << "Moving back to panel" <<"\n";
  Robot::IHM_Panel();
}

void Robot::change_speed(void) {
  std::cout << "Insert the new speed value: ";
  std::cin >> speed.data;
  speed.data = abs(speed.data);
  speed_neg.data = -speed.data;
  std::cout << "Moving back to panel" <<"\n";
  Robot::IHM_Panel();
}

void Robot::change_time(void) {
  std::cout << "Insert new duration time of movement: ";
  std::cin >> dur_timer;
  dur_timer = abs(dur_timer);
  std::cout << "Moving back to panel" <<"\n";
  Robot::IHM_Panel();
}

void Robot::detect_obs(void) {
  float largura = 0;
  float soma = 0;
  float dist;
  while ((largura*1000) < 30){
    soma = 0;
    ROS_INFO("Obstacle not detected, proceding!");
    ros::getGlobalCallbackQueue()->callAvailable(ros::WallDuration(0));
    for (int pos = 0; pos < hokuyo_info.size() ; pos++){
      soma = soma + hokuyo_info[pos];
    }
    largura = hokuyo_inc_angle.data*soma;
    ROS_INFO("largura lida em mm: %.2f", largura*1000);
    traction_1c_path.publish(speed);
    traction_1f_path.publish(speed_neg);
    traction_2f_path.publish(speed);
    traction_1b_path.publish(speed_neg);
    traction_2b_path.publish(speed);
  }
  ROS_INFO("Obstacle detected, preparing to aprox. it");
  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);
  dist = sqrt(pow(soma/hokuyo_info.size(),2)-pow(0.4,2));
  Robot::aprox_obs(dist);
  std::cout << "Finished all movements, Moving back to panel" <<"\n";
  Robot::IHM_Panel();
}

void Robot::aprox_obs(float dist) {
  std_msgs::Float64 speed_aprox;
  std_msgs::Float64 speed_neg_aprox;
  float timer_aprox;
  double current_time, starting_time, actual_time;

  speed_aprox.data = 5;
  speed_neg_aprox.data = -speed_aprox.data;
  timer_aprox = ((dist-0.08)/speed_aprox.data)*60;

  traction_1c_path.publish(speed_aprox);
  traction_1f_path.publish(speed_neg_aprox);
  traction_2f_path.publish(speed_aprox);
  traction_1b_path.publish(speed_neg_aprox);
  traction_2b_path.publish(speed_aprox);

  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < timer_aprox){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);
  ROS_INFO("Aprox. completed, preparing to surpass it");
  Robot::surpass_obs();
}

void Robot::surpass_obs(void) {
  float surpass_timer;
  double current_time, starting_time, actual_time;
  float dist;
  std_msgs::Float64 speed_surpass;
  std_msgs::Float64 speed_neg_surpass;
  std_msgs::Float64 auxiliar_var_1,auxiliar_var_2;

  //move up front arm to allow claws to rotate
  auxiliar_var_1.data = -0.3;
  auxiliar_var_2.data = 0.1;
  joint_1f_path.publish(auxiliar_var_1);
  joint_2f_path.publish(auxiliar_var_2);

  //Create a delay, to allow the arm to go up before rotating claws
  surpass_timer = 1;
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  //Rotate claws
  auxiliar_var_1.data = 3.14;
  auxiliar_var_2.data = 3.14;
  claw_1f_path.publish(auxiliar_var_1);
  claw_2f_path.publish(auxiliar_var_2);

  //Move some distance to allow front arm to pass the obstacle
  dist = 0.6;
  speed_surpass.data = 5;
  speed_neg_surpass.data = -speed_surpass.data;
  surpass_timer = ((dist-0.08)/speed_surpass.data)*60;

  traction_1c_path.publish(speed_surpass);
  traction_1f_path.publish(speed_neg_surpass);
  traction_2f_path.publish(speed_surpass);
  traction_1b_path.publish(speed_neg_surpass);
  traction_2b_path.publish(speed_surpass);

  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);

  //rotate the claws back to original position
  auxiliar_var_1.data = 0;
  auxiliar_var_2.data = 0;
  claw_1f_path.publish(auxiliar_var_1);
  claw_2f_path.publish(auxiliar_var_2);

  //small delay to give sufficient time to claws rotate back to position
  surpass_timer = 0.5;
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  //move front arm back to original position
  auxiliar_var_1.data = 0;
  auxiliar_var_2.data = 0;
  joint_1f_path.publish(auxiliar_var_1);
  joint_2f_path.publish(auxiliar_var_2);

  //First arm passed Obstacle

  //move both arms down to force the center up
  auxiliar_var_1.data = 1.3;
  auxiliar_var_2.data = 1.3;
  joint_1f_path.publish(auxiliar_var_1);
  joint_1b_path.publish(auxiliar_var_2);

  //Create a delay, to allow the center to go up before rotating claw
  surpass_timer = 1;
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }
  //Rotate center claw
  auxiliar_var_1.data = 3.14;
  claw_1c_path.publish(auxiliar_var_1);

  //Move some distance to allow the center to pass the obstacle
  surpass_timer = ((dist-0.08)/speed_surpass.data)*60;

  traction_1c_path.publish(speed_surpass);
  traction_1f_path.publish(speed_neg_surpass);
  traction_2f_path.publish(speed_surpass);
  traction_1b_path.publish(speed_neg_surpass);
  traction_2b_path.publish(speed_surpass);

  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);

  //Rotate the center claw back to orginal position and arms back to positio to lower the cente
  auxiliar_var_1.data = 0;
  claw_1c_path.publish(auxiliar_var_1);
  joint_1f_path.publish(auxiliar_var_1);
  joint_1b_path.publish(auxiliar_var_1);

  //Central part passed Obstacle

  //move up back arm to allow claws to rotate
  auxiliar_var_1.data = -0.3;
  auxiliar_var_2.data = 0.1;
  joint_1b_path.publish(auxiliar_var_1);
  joint_2b_path.publish(auxiliar_var_2);

  //Create a delay, to allow the arm to go up before rotating claws
  surpass_timer = 1;
  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  //Rotate claws
  auxiliar_var_1.data = 3.14;
  auxiliar_var_2.data = 3.14;
  claw_1b_path.publish(auxiliar_var_1);
  claw_2b_path.publish(auxiliar_var_2);


  //Move some distance to allow front arm to pass the obstacle
  surpass_timer = ((dist-0.08)/speed_surpass.data)*60;

  traction_1c_path.publish(speed_surpass);
  traction_1f_path.publish(speed_neg_surpass);
  traction_2f_path.publish(speed_surpass);
  traction_1b_path.publish(speed_neg_surpass);
  traction_2b_path.publish(speed_surpass);

  current_time = 0;
  starting_time = ros::Time::now().toSec();
  while (current_time < surpass_timer){
    actual_time = ros::Time::now().toSec();
    current_time = actual_time - starting_time;
  }

  traction_1c_path.publish(stop_speed);
  traction_1f_path.publish(stop_speed);
  traction_2f_path.publish(stop_speed);
  traction_1b_path.publish(stop_speed);
  traction_2b_path.publish(stop_speed);

  //move claws and arm back to original position
  auxiliar_var_1.data = 0;
  claw_1b_path.publish(auxiliar_var_1);
  claw_2b_path.publish(auxiliar_var_1);
  joint_1b_path.publish(auxiliar_var_1);
  joint_2b_path.publish(auxiliar_var_1);

  ROS_INFO("Surpassed!");
}

void Robot::callback_hokuyo(const sensor_msgs::LaserScan::ConstPtr& data){

  hokuyo_info_no_treatment = data->ranges;
  hokuyo_info.clear();
  for(int i=0 ; i < hokuyo_info_no_treatment.size() ; i++){
    if (!(std::isinf(hokuyo_info_no_treatment[i]))){
      hokuyo_info.push_back(hokuyo_info_no_treatment[i]);
    }
  }
  // for(int i=0 ; i < hokuyo_info.size() ; i++){
  //   std::cout << hokuyo_info[i];
  // }
  // std::cout << "\n"<< "\n"<< "\n"<< "\n";
  // ROS_INFO("I heard: Enter");
  hokuyo_inc_angle.data = data->angle_increment;


}

void callback_hokuyo2(const sensor_msgs::LaserScan::ConstPtr& data){
  ROS_INFO("I heard: Enter");
}

int main(int argc, char **argv) {
  ros::init(argc, argv, "Piro3_move");
  Robot piro3;


  // ros::NodeHandle n;
  // ros::Subscriber hokuyo_path2;
  // hokuyo_path2 = n.subscribe<sensor_msgs::LaserScan>("/robot/scan", 1000, callback_hokuyo2);
  // ros::Rate rate(10.0);
  // while(ros::ok()){
  //   ros::spinOnce(); // this is where the magic happens!!
  //   rate.sleep();
  // }

  return 0;
}
